WordPress/gutenberg
# Font Face: to generate and print font-face styles for theme.json fonts

## What?


Introduce Font Face processing to generate and print `@font-face` styles for fonts in theme.json merged data.

It's minimalistic in what it publicly exposes to avoid BC concerns.

There's no register, deregister, or enqueue of fonts or providers.

The fonts to be processed come directly from theme.json merged data which contains the theme's defined fonts and user "activated" fonts (from the Font Library).

## Why?


See #51769.

## How?


It adds 2 new classes `WP_Font_Face` and `WP_Font_Face_Resolver` which currently live in `/lib/experimental/fonts/` folder. That folder could also house the new Fonts Library.

Jobs:
* The Resolver's job is to handle getting and formatting the font data from theme.json.
* The `WP_Font_Face`'s job is to generate and print the `@font-face` styles.
* The local provider functionality is integrated into `WP_Font_Face`. 

If in the future support for custom providers is needed, the local provider can be removed from the`WP_Font_Face` and reintroduced as a separate class without breaking BC (backwards-compatibility).

When it's time to print, `wp_print_font_faces()` handles a call to the `WP_Font_Face_Resolver::get_fonts_from_theme_json()` and passes the fonts to `WP_Font_Face::print()`.

This also means other sources could pass fonts to `WP_Font_Face::print()`, such as plugins on a classic site.

### Prepared to be merged _before_ Fonts Library

The new files will not be loaded into memory unless the Fonts Library exists. In this way, this PR can be merged into `trunk` before the Fonts Library without breaking websites that are using the Fonts API.

## TODO

- [X] Add `WP_Font_Face_Resolver`.
- [X] Add `WP_Font_Face`.
- [X] Integrate local provider into `WP_Font_Face`.
- [X] Only expose `print()` to hide the internal workings (eliminate future BC concerns).
- [X] Conditional load only when the Fonts Library exists.
- [x] Add PHPUnit tests.

## Testing Instructions

1. Enable the loading of this PR's files by:
    a. Open `lib/load.php`.
    b. Scroll down to where the Font Face files are loaded.
    c. Add `true || ` in the `if`:
```php
if ( true || class_exists( 'WP_Fonts_Library' ) || class_exists( 'WP_Fonts_Library_Controller' ) ) {
```
2. With TT3:
  a. In Site Editor using Dev Tools, check that the `<script id="wp-fonts-local">` and `@font-face` styles are present in `<head>` and the iframed editor.
  b. In a post or page, check that the `<script id="wp-fonts-local">` and `@font-face` styles are present in `<head>` and the iframed editor.
  c. On the front-end, check that the `<script id="wp-fonts-local">` and `@font-face` styles are present.

```html
<style id="wp-fonts-local">
@font-face{font-family:"DM Sans";font-style:normal;font-weight:400;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/dm-sans/DMSans-Regular.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"DM Sans";font-style:italic;font-weight:400;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/dm-sans/DMSans-Regular-Italic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"DM Sans";font-style:normal;font-weight:700;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/dm-sans/DMSans-Bold.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"DM Sans";font-style:italic;font-weight:700;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/dm-sans/DMSans-Bold-Italic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:normal;font-weight:300;font-display:block;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/ibm-plex-mono/IBMPlexMono-Light.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:normal;font-weight:400;font-display:block;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/ibm-plex-mono/IBMPlexMono-Regular.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:italic;font-weight:400;font-display:block;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/ibm-plex-mono/IBMPlexMono-Italic.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"IBM Plex Mono";font-style:normal;font-weight:700;font-display:block;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/ibm-plex-mono/IBMPlexMono-Bold.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:Inter;font-style:normal;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/inter/Inter-VariableFont_slnt,wght.ttf') format('truetype');font-stretch:normal;}
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"Source Serif Pro";font-style:italic;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentythree/assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2') format('woff2');font-stretch:normal;}

</style>
```
2. With TT1:
    * In the `functions.php`, add the following code:
```php
add_action( 'wp_head', 'tt1_test_print_font_faces', 50 );
add_action( 'admin_print_styles', 'tt1_test_print_font_faces', 50 );
function tt1_test_print_font_faces() {
	$fonts = array(
			'Source Serif Pro' => array(
					array(
							'font-family' => 'Source Serif Pro',
							'font-style'   => 'normal',
							'font-weight'  => '200 900',
							'src'          => get_theme_file_uri( 'assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2' ),
					),
					array(
							'font-family' => 'Source Serif Pro',
							'font-style'   => 'italic',
							'font-weight'  => '200 900',
							'src'          => get_theme_file_uri( 'assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2' ),
					),
			),
	);
	wp_print_font_faces( $fonts );
}
```
   * Using Dev Tools, check in the front-end and admin area that the `<script id="wp-fonts-local">` and `@font-face` styles are present.

```html
<style id="wp-fonts-local">
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentyone/assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');}
@font-face{font-family:"Source Serif Pro";font-style:italic;font-weight:200 900;font-display:fallback;src:url('http://localhost:8889/wp-content/themes/twentytwentyone/assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2') format('woff2');}

</style>
```
